Splat pop

azodichr opened the pull request:
Updated pull request following Luke's instructions for submitting a PR from Luke's splatPop branch.

lazappi commented:
I restarted the GHA (they failed originally for some unrelated reason). Looks like there are some errors/warnings around the VCF datasets https://github.com/Oshlack/splatter/pull/106/checks?check_run_id=1219314609. Can you have a look at those?

lazappi commented on R/splatPop-simulate.R:
```diff
@@ -0,0 +1,849 @@
+#'     \item Add quantile normalized gene mean and cv info the eQTL key.
+#'     \item Add eQTL effects to means matrix.}
+#'
+#' @return A list containing: `means` a data.frame (or list of dataframes if
+#' n.groups > 1) with the simulated mean gene expression value for each gene
+#' (row) and each sample (column) and `key` a data.frame with population 
+#' information including eQTL and group effects.
```
I wonder if we can store these in `SplatPopParams` and return that? This is what I did for `kersplat` but it was a bit fiddly to get it working...

azodichr replied:
Done!

lazappi commented on R/splatPop-simulate.R:
```diff
@@ -0,0 +1,849 @@
+#' @importFrom SummarizedExperiment rowData rowData<-
+#' @export
+#' 
+splatPopSimulateSC <- function(sim_means,
```
I wonder if there might be better names for this and `splatPopSimulateMeans`? I used `kersplatSetup()` and `kersplatSample()` but maybe not appropriate here. I'll have a think but not super important.

azodichr replied:
I went with these names because it clearly distinguishes between the two phases of simulations in splatPop. Happy to hear other suggestions!

lazappi commented on R/splatPopParams-methods.R:
```diff
@@ -0,0 +1,106 @@
+    return(object)
+})
+
+#' @rdname setParams
+setMethod("setParams", "splatPopParams", function(object, update = NULL, ...) {
+    
+    checkmate::assertClass(object, classes = "splatPopParams")
+    checkmate::assertList(update, null.ok = TRUE)
+    
+    update <- c(update, list(...))
+    
+    object <- callNextMethod(object, update)
+    
+    return(object)
+})
```
There's nothing specific to `SplatPopParams` here so I think we can drop it and just use the `SplatParams` method.

azodichr replied:
does that mean modifying setMethod in SplatParams-methods.R to: 
setMethod("setParams", c("splatPop", "splatPopParams"), function(object, update = NULL, ...)

lazappi replied:
No I think you should just be able to delete this and inheritance should fall back to the `SplatParams` method without any modifications. I think so anyway... Should be easy enough to test by commenting this out, loading the package and seeing if `setParams()` still works.

azodichr replied:
That worked so I will remove setParams from splatPopParams-methods.

lazappi commented on R/utils.R:
```diff
@@ -1,3 +1,7 @@
+#' @importFrom utils globalVariables
+utils::globalVariables(c("pval_nominal", 'perc', 'gene_id',  'type', 'eQTL',
+                         'chr', 'gene_mid', "maf", "data", "vcf", "gff", "dna"))
+
```
What are these variable used for? I'd prefer to avoid any global variables if we can.

azodichr replied:
I did this to remove the warnings that look like: 

`no visible binding for global variable ‘maf’`

If you have a preferred solution, I'm happy to change it!

lazappi replied:
Ugh, I know that warning well, it's the worst. Do you know where that is coming from? There are a few solutions depending on the cause. If it's not obvious I can probably track it down.

azodichr replied:
Three (vcf, gff, dna) I removed because they were related to the vcfR package. The others are all data.frame column names.

lazappi replied:
I think when the column already exists using `$` is fine but it complains if you use it to create a new column. You should be able to fix it by using `df[, "col"]` or `df[["col"]]` instead of `df$col`.

azodichr replied:
Cool! Fixed :)

lazappi commented on tests/testthat/test-splatPop-simulate.R:
```diff
@@ -0,0 +1,35 @@
+context("splatPopSimulate")
+
+library(vcfR)
```
I don't think we need this library call

azodichr replied:
Without loading the package I get the following error during testing:

```
test-splatPop-simulate.R:8: warning: (unknown)
data set 'vcfR_example' not found

test-splatPop-simulate.R:8: error: (unknown)
object 'vcf' not found
```

lazappi replied:
This might be from the `data(vcfR_example)` call, you probably need `data(vcfR_example, package = "vcfR")`.

azodichr replied:
This is still required for the tests to run (although now I load VariantAnnotation)

lazappi reviewed (commented):
Ok. I've had a look through most of the code but not the vignettes yet. It's mostly minor things, the only potentially big one is whether we can store more things in `SplatPopParams` but we can discuss further whether or not that is a good idea.

lazappi commented:
Some notes on style. Mostly we try and follow the Bioconductor style guide https://www.bioconductor.org/developers/how-to/coding-style/. It's up to you whether you want to do these as you go or I can fix them all at the end.

Documentation:
* We aren't using the Roxygen markdown mode so things like `** **` won't work in the function docs. Need to use the old style tags instead (`\code{ }`).
* Don't need to include defaults for function arguments in the docs because these are shown on the help pages anyway.

Code:
* Have spaces around all operators, including in function calls (e.g. `f(a = 1)` rather than `f(a=1)`).
* Functions are named using `camelCase`
* Variables are named using `dot.separation`

lazappi commented on R/splatPop-simulate.R:
```diff
@@ -145,7 +145,7 @@ splatPopSimulateMeans <- function(vcf=mock_vcf(),
     set.seed(getParam(params, "seed"))
     
-    nGroups <- getParam(params, "nGroups")
+    nGroups <- length(getParam(params, "group.prob"))
```
Why did `nGroups` not work? They should be the same thing... Just wondering if there is a bigger issue here we might need to look at.

azodichr replied:
Groups wasn't working for me because I hadn't included all of your checks to setMethod("setParam"), so updating group.prob wasn't automatically changing the nGroup. I have fixed that now and reverted this code back to calling nGroups for simplicity.

lazappi commented on R/splatPop-simulate.R:
```diff
@@ -50,7 +49,7 @@
                              params = newSplatPopParams(nGenes=1000),
                              method = c("single", "groups", "paths"),
-                             gff = FALSE,
+                             gff = NA,
```
Maybe `NULL` is better than `NA`?

lazappi commented on R/splatPop-simulate.R:
```diff
@@ -459,17 +459,15 @@ splatPopParseVCF <- function(vcf, params){
     nGenes <- getParam(params, "nGenes")
     
-    if(random.genes){
+    if(is.logical(gff)){
```
You probably want `is.na()` here (or `is.null()`)

azodichr replied:
I've removed random.genes from the splatPopParam object and have just made the function determine default to random if genes aren't provided in a key or a gff.

azodichr commented on DESCRIPTION:
```diff
@@ -46,6 +47,7 @@ Suggests:
     covr,
     cowplot,
     knitr,
+    magick,
```
Yes, that can be removed!

azodichr replied:
I take that back, I get this error during R CMD when magick is removed: 
`The magick package is required to crop "/Users/cazodi/GitHub/splatter.Rcheck/vign_test/splatter/vignettes/splat_params_files/figure-html/batch-factors-1.png" but not available.`

I have left it in for now.

azodichr commented on DESCRIPTION:
```diff
@@ -34,6 +34,7 @@ Imports:
     locfit,
     matrixStats,
     methods,
+    preprocessCore,
```
I've moved preprocessCore to suggested and added requireNamespace to the quantile normalization function.

azodichr replied:
I am getting the following warning during R CMD: 

```
> checking for unstated dependencies in examples ... WARNING
  'library' or 'require' calls not declared from:
    ‘c’ ‘preprocessCore’ ‘vcfR’
```
I tried to set these up like in your ZINB function since that also has a dependency suggestion, but I must have missed something.

lazappi replied:
I don't think you can give a vector to `requireNamespace()` so you might have to do something like:

```r
if (requireNamespace("preprocessCore", quietly = TRUE) & requireNamespace("vcfR", quietly = TRUE)) {
...
}
```

azodichr commented on R/mock-data.R:
```diff
@@ -0,0 +1,134 @@
+    data(vcfR_example)
+    mock_vcf <- vcf
+    rm(gff, vcf, dna, pos=1)
+    mock_vcf@meta[2] <- "##source=\"Mock\""    
```
There are not functions besides @ for extracting components from vcfR. There are a few bioconductor packages that have vcf objects (VariantAnnotation, TVTB, and VCFArray), however none of them have a "mock" function and none come with an example vcf object. So the only way to create a vcf object with any of those three software is to provide a vcf file yourself. I went with vcfR specifically because it already had the example data.  Happy to chat more about this. The data loaded in vcfR_example is just 3.4 MB.

lazappi replied:
When I asked some Bioconductor people the general consensus was to use **VariantAnnotation**. I think it includes an example dataset. Once it is installed you should be able to do something like:

```r
 fl <- system.file("extdata", "chr22.vcf.gz", package="VariantAnnotation")
vcf <- readVcf(fl, "hg19")
```

Do you think it would be possible to use that instead? It might make things easier for people if we use a more standard format.

Alternatively how much of the VCF do you actually need? I'm wondering if it might be easier to just store a `data.frame` and extract the information for a VCF during estimation. I realise that's probably closer to what you had originally. I don't trust users to do the VCF to `data.frame` conversion themselves though so we will still have to deal with VCFs so maybe that doesn't make things any easier...

Sorry to make you look at this again but I think things will break if we need to change it later so I'd like to get it right now before anyone is using it.

azodichr replied:
vcfR is replaced with VariantAnnotation

azodichr commented on R/mock-data.R:
```diff
@@ -0,0 +1,134 @@
+mock_vcf <- function(n_snps=1e4, n_samples=10, chromosome=1){
+    
+    if (requireNamespace("vcfR", quietly = TRUE))
+    data(vcfR_example)
```
See comment below

azodichr replied:
see comment below.

azodichr replied:
I did change this to data(vcfR_example, envir = environment()) so that the example vcf data is loaded into the evaluation environment and won't take up space or replace existing objects in user's environments. Hoping that solves the issues!

azodichr commented on R/splatPop-estimate.R:
```diff
@@ -0,0 +1,176 @@
+#' @param params splatPopParams object containing parameters for the 
+#'         simulation of the mean expression levels for the population.
+#'        See \code{\link{splatPopParams}} for details.
+#' @param means Dataframe of real gene means across a population, where 
+#'        each row is a gene and each column is an individual in the population.
```
I have updated this and the mock_bulk_matrix function to use a matrix instead of a data.frame.

azodichr commented on R/splatPop-estimate.R:
```diff
@@ -0,0 +1,176 @@
+                           stringsAsFactors=FALSE)
+
+    for(b in names(bins)){
+        stst <- unlist(strsplit(gsub("\\)|\\(|\\[|\\]", "", b), ","))
```
I changed the variable name from stst (start_stop) to min_max to make it more clear what that line is doing. I also made a new variable called re.brack_paren and set that equal to "\\[|\\]|\\)|\\(" so that it is clear that gsub call is just removing special characters.

azodichr commented on R/splatPop-simulate.R:
```diff
@@ -0,0 +1,849 @@
+#' Simulate scRNA-seq count data using the splat model for a population of
+#' individuals with correlation structure.
+#'
+#' @param vcf vcfR object containing genotypes of samples to simulate.
```
I was hesitant to do that because vcf objects can be very large. I get wanting to be able to recreate the whole simulation from the params object, but I've kept the gff, key, and vcf objects out of the param object for size concerns. If you don't think we should worry about its size, then I can include vcf or all three :)

lazappi replied:
Maybe it's worthing checking how big each of these are (you can use `object.size()`)? I'm wondering how big they are compared to the expression matrices? That might help us decide if size is an issue or not.

azodichr replied:
Haven't been able to come up with a good way to add info about the vcf to the params object automatically. My only thought is to have a vcf and gff container in the params object and give people the option to set it equal to a character, then in the vignette just encourage folks to put a helpful descriptor of the file or a URL to where it was downloaded from.

azodichr commented on R/splatPop-simulate.R:
```diff
@@ -0,0 +1,849 @@
+#'        (eg. cell types), "paths" which selects cells from continuous
+#'        trajectories (eg. differentiation processes).
+#' @param gff Either FALSE (if random.genes=TRUE) or a data.frame object 
+#'        containing a GFF/GTF file.
```
Again, I kept it out because of size concerns. Happy to discuss further!

azodichr commented on R/splatPop-simulate.R:
```diff
@@ -0,0 +1,849 @@
+#'        trajectories (eg. differentiation processes).
+#' @param gff Either FALSE (if random.genes=TRUE) or a data.frame object 
+#'        containing a GFF/GTF file.
+#' @param key Either FALSE or a data.frame object containing a full or partial
+#'        splatPop key.
```
The key is one of the 3 objects returned by splatPopSimulate, it contains information about the genes simulated, including mean/cv, eQTL assignments, DE, etc. The reason it is also an optional input, is that you could provide splatPopSimulateMeans with the key and it would simulate a new population using the same parameters. Or if you had real genes you want to simulate SC expression for, you can provide the genes their location and their known bulk expression levels, and let splatPopSimulate fill in the rest. 

It is a bit clunky, but this feature was strongly requested by one of our collaborators.

lazappi replied:
👍 That sounds like a reasonable feature, still think we can come up with a better name though 😼? I think we can have this kind of optional things as part of `splatPopParams` but depends how hard it would be to implement (and the size issue).

azodichr replied:
I've kept the splatPop key object as an output of splatPopSimulateMeans, but now, when you run splatPopSimulate the information from that key is now stored in the sce object in the rowData (since they key contains info for each gene).

azodichr commented on R/splatPop-simulate.R:
```diff
@@ -0,0 +1,849 @@
+#' @seealso
+#' \code{\link{splatPopSimulateMeans}}, \code{\link{splatPopSimulateSC}}
+#'
+#' @return A list containing: `means` a dataframe (or list of dataframes if
+#' n.groups > 1) with the simulated mean gene expression value for each gene
+#' (row) and each sample (column), `key` a dataframe with population information
+#' including eQTL and group effects, and `sce` a SingleCellExperiment object
+#' containing simulated counts and intermediate values.
```
I hadn't thought about doing that! I think that sounds like a more streamlined way to hold the data, but I'm going to run that by Davis tomorrow and see if he can come up with any good reason not to!

azodichr replied:
Now all of this is in the sce object!

azodichr commented on R/params-functions.R:
```diff
@@ -12,7 +12,7 @@
 getParams <- function(params, names) {
 
-    checkmate::assertClass(params, classes = "Params")
+    #checkmate::assertClass(params, classes = "Params")
```
I honestly don't remember exactly why I commented these out! I think it was because of issues I was having early on in splatPop development that I've since fixed. I've added them back and so far no problems...

azodichr commented:
Thanks for your comments, Luke! Sorry it took so long to reply, I didn't get notified when comments were added, so I just saw them today. I'll keep a closer eye on the branch. I've addressed all your comments except those regarding the changes to what goes in the splatPopParams object and what goes in the final sce object. I'll discuss those things with Davis and start making changes after we have a plan! I did my best with all the bioconductor best practices changes, but please do point out things I've missed.

lazappi commented:
No worries! I'm sure Davis will have some useful comments so please discuss with him.

I tried to close the conversations but I wasn't sure if you had done some of them yet. If you think you have addressed something feel free to mark it as resolved.

azodichr commented:
I've pushed a new update with some of these minor changes and all the output now contained in the sce object. Still working on a few changes, including switching to VariantAnnotation and incorporating info about the gff and vcf into the params object. I don't want to include the whole data set in the params because vcf files can easily be in the GBs, but it would be good to automatically add some identifying info about those datasets to params.

azodichr commented on R/mock-data.R:
```diff
@@ -0,0 +1,134 @@
+#' 
+mock_vcf <- function(n_snps=1e4, n_samples=10, chromosome=1){
+    
+    if (requireNamespace("vcfR", quietly = TRUE))
```
vcfR is replaced with VariantAnnotation

lazappi commented on R/mock-data.R:
```diff
@@ -0,0 +1,157 @@
+                                              width = nchar(n.samples),
+                                              format = "d",
+                                              flag = "0"))
+    snp_names <- paste0("snp_", formatC(1:n.snps,
```
Change to `seq_len()`

lazappi commented on R/mock-data.R:
```diff
@@ -0,0 +1,157 @@
+        stop("Creating a mock VCF requires the 'VariantAnnotation' package.")
+    }
+
+    sample_names <- paste0("sample_", formatC(1:n.samples,
```
Change to `seq_len()`

lazappi commented on R/splat-estimate.R:
```diff
@@ -13,7 +13,8 @@
 #' \code{\link{splatEstDropout}}
 #'
-#' @return SplatParams object containing the estimated parameters.
+#' @return SplatParams or splatPopParams object containing the estimated 
```
Think it's fine to just say SplatParams here because SplatPopParams is a subclass.

lazappi commented on R/splat-estimate.R:
```diff
@@ -41,7 +42,9 @@ splatEstimate.SingleCellExperiment <- function(counts,
 splatEstimate.matrix <- function(counts, params = newSplatParams()) {
 
-    checkmate::assertClass(params, "SplatParams")
+    if(class(params) == "splatPopParams"){
+        checkmate::assertClass(params, "splatPopParams")
+    }else{checkmate::assertClass(params, "SplatParams")}
```
Check how the class check works

lazappi commented on R/splat-simulate.R:
```diff
@@ -7,9 +7,9 @@
+#'        (eg. cell types), or "paths" which selects cells from continuous
 #'        trajectories (eg. differentiation processes).
-#' @param verbose logical. Whether to print progress messages.
+#' @param verbose logical. Whether to print progress messages (Default=TRUE).
```
Remove default

lazappi commented on R/splatPop-estimate.R:
```diff
@@ -0,0 +1,178 @@
+#' real data. See the individual estimation functions for more details on
+#' how this is done.
+#'
+#' @param params splatPopParams object containing parameters for the
```
Capitalise `SplatPopParams`

lazappi commented on R/splatPop-estimate.R:
```diff
@@ -0,0 +1,178 @@
+#' Estimate rate and shape parameters for the gamma distribution used to
+#' simulate eQTL (eSNP-eGene) effect sizes.
+#'
+#' @param eqtl Data.frame with all or top eQTL pairs from a real eQTL analysis.
```
data.frame

lazappi commented on R/splatPop-estimate.R:
```diff
@@ -0,0 +1,178 @@
+
+#' Estimate gene mean and gene mean variance parameters
+#'
+#' The Shapiro-Wilks test is used to determine if the library sizes are
+#' normally distributed. If so a normal distribution is fitted to the library
+#' sizes, if not (most cases) a log-normal distribution is fitted and the
+#' estimated parameters are added to the params object. See
+#' \code{\link[fitdistrplus]{fitdist}} for details on the fitting.
```
Maybe left over from somewhere else?

lazappi commented on R/splatPop-simulate.R:
```diff
@@ -0,0 +1,849 @@
+    for (idx in seq_len(n.groups)) {
+        de.facs <- getLNormFactors(nGenes, de.prob, de.downProb,
+                                   de.facLoc, de.facScale)
+        key[, paste0(groups[idx], ".GroupEffect")] <- de.facs
```
Remove `"."` from column names

lazappi commented on R/splatPop-simulate.R:
```diff
@@ -0,0 +1,849 @@
+#'
+#' @return SingleCellExperiment with simulated sc counts.
+#'
+splatPopCleanSCE <- function(sim.all){
+
+    # Remove redundant sce info
+    keep.id <-  gsub("_Gene", "", names(rowData(sim.all))[1])
+
+    shared.out.factor <- rowData(sim.all)[[paste0(keep.id, "_OutlierFactor")]]
+    rowData(sim.all)[grepl("_OutlierFactor", names(rowData(sim.all)))] <- NULL
+    rowData(sim.all)$Shared_OutlierFactor <- shared.out.factor
+
+    rowData(sim.all)[grepl("_Gene", names(rowData(sim.all)))] <- NULL
+    metadata(sim.all)[2:length(names(metadata(sim.all)))] <- NULL
+
+    return(sim.all)
+}
```
Can probably replace this with the new `minmiseSCE()` function (which I have written but is in another branch.

lazappi commented on R/splatPopParams-methods.R:
```diff
@@ -0,0 +1,166 @@
+    
+    checkmate::assertClass(params, "splatPopParams")
+
+    for (pkg in c("VariantAnnotation", "preprocessCore")) {
+        if (!requireNamespace(pkg, quietly = TRUE)) {
+            stop("The splatPop simulation requires the ", pkg, " package.")
+        }
+    }
```
Check should probably go first

lazappi commented on R/splatPopParams-methods.R:
```diff
@@ -0,0 +1,166 @@
+
+    pp <- list("Population params:" = c("(mean.shape)" = "pop.mean.shape",
+                                        "(mean.rate)" = "pop.mean.rate",
+                                        #"[group.prop]" = "group.prop",
```
Delete comment

lazappi commented on R/splatPopParams-methods.R:
```diff
@@ -0,0 +1,166 @@
+
+
+#' @rdname setParam
+setMethod("setParam", "splatPopParams", function(object, name, value) {
```
I think most of this should be covered by the `SplatParams` method, but need to check.

lazappi commented on R/utils.R:
```diff
@@ -1,3 +1,7 @@
+#' @importFrom utils globalVariables
+utils::globalVariables(c("pval_nominal", 'perc', 'gene_id', 'type', 'eQTL',
+                         'chr', 'gene_mid', "maf")) #, "data"))
+
```
These should be able to be deleted now

lazappi commented on tests/testthat/test-splatPopParams.R:
```diff
@@ -0,0 +1,59 @@
+    expect_silent(setParams(params, eqtl.maf.max = 0.4, eqtl.n = 10))
+})
+
+### These tests are also run in test-splatParams.R, please update both
+test_that("path.from checks work", {
+    pp <- setParams(params, group.prob = c(0.5, 0.5))
+    pp <- setParamUnchecked(pp, "path.from", c(0, 1))
+    expect_silent(validObject(pp))
+    pp <- setParamUnchecked(pp, "path.from", c(0, 3))
+    expect_error(validObject(pp), "invalid class")
+    pp <- setParamUnchecked(pp, "path.from", c(1, 0))
+    expect_error(validObject(pp), "path cannot begin at itself")
+    pp <- newSplatParams()
+    pp <- setParams(pp, group.prob = c(0.3, 0.3, 0.4))
+    pp <- setParamUnchecked(pp, "path.from", c(2, 1, 1))
+    expect_error(validObject(pp), "origin must be specified in path.from")
+    pp <- setParams(params, group.prob = c(0.5, 0.5), path.from = c(0, 1))
+    expect_warning(setParam(pp, "group.prob", 1),
+                   "nGroups has changed, resetting path.from")
+    pp <- newSplatParams()
+    pp <- setParams(pp, group.prob = c(0.25, 0.25, 0.25, 0.25))
+    pp <- setParamUnchecked(pp, "path.from", c(0, 4, 2, 3))
+    expect_error(validObject(pp), "path.from cannot contain cycles")
+})
+
+### These tests are also run in test-splatParams.R, please update both
+test_that("dropout.type checks work", {
+    expect_error(setParam(params, "dropout.type", "cell"),
+                 "dropout.type cannot be set to 'cell'")
+    pp <- setParams(params, dropout.mid = rep(1, 100),
+                    dropout.shape = rep(1, 100))
+    expect_silent(setParam(pp, "dropout.type", "cell"))
+    expect_error(setParam(params, "dropout.type", "a"),
+                 "dropout.type must be one of: ")
+})
```
Not sure we need to recheck these

lazappi commented on vignettes/splatPop.Rmd:
```diff
@@ -0,0 +1,418 @@
+---
+title: "splatPop: simulating single-cell data for populations"
```
Proofread vignette

lazappi reviewed (commented):
@azodichr These are mostly minor things that I will fix so you can ignore them.

Only thing I would ask you to do before we merge is to add yourself as a contributor (`role = c("ctb")`) in the `DESCRIPTION`. You have definitely contributed enough to deserve some acknowledgement!

azodichr commented:
Great. I’ve added myself to the description. 

> On Oct 21, 2020, at 7:37 PM, Luke Zappia <notifications@github.com> wrote:
> 
> 
> @lazappi commented on this pull request.
> 
> @azodichr <https://github.com/azodichr> These are mostly minor things that I will fix so you can ignore them.
> 
> Only thing I would ask you to do before we merge is to add yourself as a contributor (role = c("ctb")) in the DESCRIPTION. You have definitely contributed enough to deserve some acknowledgement!
> 
> In R/mock-data.R <https://github.com/Oshlack/splatter/pull/106#discussion_r509076318>:
> 
> > +#' @importFrom stats setNames
> +#'
> +#' @export
> +#'
> +mockVCF <- function(n.snps = 1e4, n.samples = 10, chromosome = 22){
> +
> +
> +    if (!requireNamespace("VariantAnnotation", quietly = TRUE)) {
> +        stop("Creating a mock VCF requires the 'VariantAnnotation' package.")
> +    }
> +
> +    sample_names <- paste0("sample_", formatC(1:n.samples,
> +                                              width = nchar(n.samples),
> +                                              format = "d",
> +                                              flag = "0"))
> +    snp_names <- paste0("snp_", formatC(1:n.snps,
> Change to seq_len()
> 
> In R/mock-data.R <https://github.com/Oshlack/splatter/pull/106#discussion_r509076577>:
> 
> > +#' @param chromosome Chromosome name
> +#'
> +#' @return data.frame containing mock gff data.
> +#'
> +#' @importFrom stats setNames
> +#'
> +#' @export
> +#'
> +mockVCF <- function(n.snps = 1e4, n.samples = 10, chromosome = 22){
> +
> +
> +    if (!requireNamespace("VariantAnnotation", quietly = TRUE)) {
> +        stop("Creating a mock VCF requires the 'VariantAnnotation' package.")
> +    }
> +
> +    sample_names <- paste0("sample_", formatC(1:n.samples,
> Change to seq_len()
> 
> In R/splat-estimate.R <https://github.com/Oshlack/splatter/pull/106#discussion_r509077769>:
> 
> > @@ -13,7 +13,8 @@
>  #' \code{\link{splatEstOutlier}}, \code{\link{splatEstBCV}},
>  #' \code{\link{splatEstDropout}}
>  #'
> -#' @return SplatParams object containing the estimated parameters.
> +#' @return SplatParams or splatPopParams object containing the estimated 
> Think it's fine to just say SplatParams here because SplatPopParams is a subclass.
> 
> In R/splat-estimate.R <https://github.com/Oshlack/splatter/pull/106#discussion_r509078165>:
> 
> > +    if(class(params) == "splatPopParams"){
> +        checkmate::assertClass(params, "splatPopParams")
> +    }else{checkmate::assertClass(params, "SplatParams")}
> Check how the class check works
> 
> In R/splat-simulate.R <https://github.com/Oshlack/splatter/pull/106#discussion_r509080167>:
> 
> >  #'        trajectories (eg. differentiation processes).
> -#' @param verbose logical. Whether to print progress messages.
> +#' @param verbose logical. Whether to print progress messages (Default=TRUE).
> Remove default
> 
> In R/splatPop-estimate.R <https://github.com/Oshlack/splatter/pull/106#discussion_r509080740>:
> 
> > @@ -0,0 +1,178 @@
> +#' Estimate population/eQTL simulation parameters
> +#'
> +#' Estimate simulation parameters for the eQTL population simulation from
> +#' real data. See the individual estimation functions for more details on
> +#' how this is done.
> +#'
> +#' @param params splatPopParams object containing parameters for the
> Capitalise SplatPopParams
> 
> In R/splatPop-estimate.R <https://github.com/Oshlack/splatter/pull/106#discussion_r509081212>:
> 
> > +
> +    # Get parameters for population wide gene mean and variance distributions
> +    if (!is.null(means)) {
> +        params <- splatPopEstimateMeanCV(params, means)
> +    }
> +
> +    return(params)
> +
> +}
> +
> +#' Estimate eQTL Effect Size parameters
> +#'
> +#' Estimate rate and shape parameters for the gamma distribution used to
> +#' simulate eQTL (eSNP-eGene) effect sizes.
> +#'
> +#' @param eqtl Data.frame with all or top eQTL pairs from a real eQTL analysis.
> data.frame
> 
> In R/splatPop-estimate.R <https://github.com/Oshlack/splatter/pull/106#discussion_r509082071>:
> 
> > +#' The Shapiro-Wilks test is used to determine if the library sizes are
> +#' normally distributed. If so a normal distribution is fitted to the library
> +#' sizes, if not (most cases) a log-normal distribution is fitted and the
> +#' estimated parameters are added to the params object. See
> +#' \code{\link[fitdistrplus]{fitdist}} for details on the fitting.
> Maybe left over from somewhere else?
> 
> In R/splatPop-simulate.R <https://github.com/Oshlack/splatter/pull/106#discussion_r509085074>:
> 
> > +        glob.genes <- subset(key, key$eQTL.type == "global")$geneID
> +        g.specific <- sample(glob.genes, size = n.specific.each)
> +        key[["eQTL.type"]][key$geneID %in% g.specific] <- g
> +    }
> +
> +    # Assign group-specific effects (differential expression, not eQTL)
> +    nGenes <- nrow(key)
> +    de.prob <- getParam(params, "de.prob")
> +    de.downProb <- getParam(params, "de.downProb")
> +    de.facLoc <- getParam(params, "de.facLoc")
> +    de.facScale <- getParam(params, "de.facScale")
> +
> +    for (idx in seq_len(n.groups)) {
> +        de.facs <- getLNormFactors(nGenes, de.prob, de.downProb,
> +                                   de.facLoc, de.facScale)
> +        key[, paste0(groups[idx], ".GroupEffect")] <- de.facs
> Remove "." from column names
> 
> In R/splatPop-simulate.R <https://github.com/Oshlack/splatter/pull/106#discussion_r509086243>:
> 
> > +splatPopCleanSCE <- function(sim.all){
> +
> +    # Remove redundant sce info
> +    keep.id <-  gsub("_Gene", "", names(rowData(sim.all))[1])
> +
> +    shared.out.factor <- rowData(sim.all)[[paste0(keep.id, "_OutlierFactor")]]
> +    rowData(sim.all)[grepl("_OutlierFactor", names(rowData(sim.all)))] <- NULL
> +    rowData(sim.all)$Shared_OutlierFactor <- shared.out.factor
> +
> +    rowData(sim.all)[grepl("_Gene", names(rowData(sim.all)))] <- NULL
> +    metadata(sim.all)[2:length(names(metadata(sim.all)))] <- NULL
> +
> +    return(sim.all)
> +}
> Can probably replace this with the new minmiseSCE() function (which I have written but is in another branch.
> 
> In R/splatPopParams-methods.R <https://github.com/Oshlack/splatter/pull/106#discussion_r509086662>:
> 
> > +    for (pkg in c("VariantAnnotation", "preprocessCore")) {
> +        if (!requireNamespace(pkg, quietly = TRUE)) {
> +            stop("The splatPop simulation requires the ", pkg, " package.")
> +        }
> +    }
> Check should probably go first
> 
> In R/splatPopParams-methods.R <https://github.com/Oshlack/splatter/pull/106#discussion_r509086993>:
> 
> > +        valid <- TRUE
> +    } else {
> +        valid <- checks[checks != TRUE]
> +        valid <- paste(names(valid), valid, sep = ": ")
> +    }
> +
> +    return(valid)
> +})
> +
> +
> +#' @importFrom methods callNextMethod
> +setMethod("show", "splatPopParams", function(object) {
> +
> +    pp <- list("Population params:" = c("(mean.shape)" = "pop.mean.shape",
> +                                        "(mean.rate)" = "pop.mean.rate",
> +                                        #"[group.prop]" = "group.prop",
> Delete comment
> 
> In R/splatPopParams-methods.R <https://github.com/Oshlack/splatter/pull/106#discussion_r509087705>:
> 
> > +                                        "(cv.params)" = "pop.cv.param"),
> +               "eQTL params:" = c("[eqtl.n]"    = "eqtl.n",
> +                                  "[eqtl.dist]" = "eqtl.dist",
> +                                  "[eqtl.maf.min]" = "eqtl.maf.min",
> +                                  "[eqtl.maf.max]" = "eqtl.maf.max",
> +                                  "[eqtl.group.specific]" = "eqtl.group.specific",
> +                                  "(eqtl.ES.shape)" = "eqtl.ES.shape",
> +                                  "(eqtl.ES.rate)" = "eqtl.ES.rate"))
> +
> +    callNextMethod()
> +    showPP(object, pp)
> +})
> +
> +
> +#' @rdname setParam
> +setMethod("setParam", "splatPopParams", function(object, name, value) {
> I think most of this should be covered by the SplatParams method, but need to check.
> 
> In R/utils.R <https://github.com/Oshlack/splatter/pull/106#discussion_r509088265>:
> 
> > +#' @importFrom utils globalVariables
> +utils::globalVariables(c("pval_nominal", 'perc', 'gene_id', 'type', 'eQTL',
> +                         'chr', 'gene_mid', "maf")) #, "data"))
> +
> These should be able to be deleted now
> 
> In tests/testthat/test-splatPopParams.R <https://github.com/Oshlack/splatter/pull/106#discussion_r509089536>:
> 
> > +### These tests are also run in test-splatParams.R, please update both
> +test_that("path.from checks work", {
> +    pp <- setParams(params, group.prob = c(0.5, 0.5))
> +    pp <- setParamUnchecked(pp, "path.from", c(0, 1))
> +    expect_silent(validObject(pp))
> +    pp <- setParamUnchecked(pp, "path.from", c(0, 3))
> +    expect_error(validObject(pp), "invalid class")
> +    pp <- setParamUnchecked(pp, "path.from", c(1, 0))
> +    expect_error(validObject(pp), "path cannot begin at itself")
> +    pp <- newSplatParams()
> +    pp <- setParams(pp, group.prob = c(0.3, 0.3, 0.4))
> +    pp <- setParamUnchecked(pp, "path.from", c(2, 1, 1))
> +    expect_error(validObject(pp), "origin must be specified in path.from")
> +    pp <- setParams(params, group.prob = c(0.5, 0.5), path.from = c(0, 1))
> +    expect_warning(setParam(pp, "group.prob", 1),
> +                   "nGroups has changed, resetting path.from")
> +    pp <- newSplatParams()
> +    pp <- setParams(pp, group.prob = c(0.25, 0.25, 0.25, 0.25))
> +    pp <- setParamUnchecked(pp, "path.from", c(0, 4, 2, 3))
> +    expect_error(validObject(pp), "path.from cannot contain cycles")
> +})
> +
> +### These tests are also run in test-splatParams.R, please update both
> +test_that("dropout.type checks work", {
> +    expect_error(setParam(params, "dropout.type", "cell"),
> +                 "dropout.type cannot be set to 'cell'")
> +    pp <- setParams(params, dropout.mid = rep(1, 100),
> +                    dropout.shape = rep(1, 100))
> +    expect_silent(setParam(pp, "dropout.type", "cell"))
> +    expect_error(setParam(params, "dropout.type", "a"),
> +                 "dropout.type must be one of: ")
> +})
> Not sure we need to recheck these
> 
> In vignettes/splatPop.Rmd <https://github.com/Oshlack/splatter/pull/106#discussion_r509090422>:
> 
> > @@ -0,0 +1,418 @@
> +---
> +title: "splatPop: simulating single-cell data for populations"
> Proofread vignette
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/Oshlack/splatter/pull/106#pullrequestreview-513414749>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ACZEAFQYVS2HLLFEJJ6OR5DSL2MUZANCNFSM4RYESU5A>.
>
